### PR TITLE
feat: change go.#Test.package -> packages and can be string or [...string]

### PR DESCRIPTION
**Pull request by dolanor**
The idea is to have `go.#Test.packages` that can accept either a string, or an array of string so we can do

```cue
go.#Test & {
	packages: "./..."
}
```

```cue
go.#Test & {
	packages: ["./pkg1", "./pkg2"]
}
```

## Pros
- match the `go test` better (we can use multiple packages)

## Cons
- can confuse the user (is it a string? is it an array?)

### CODE REVIEW

**Review by TomChv — APPROVED** [review]
I agree with your change but do not forget to update tests : https://github.com/dagger/dagger/runs/6038096626?check_suite_focus=true#step:7:735

**Comment by dolanor** [comment]
If this new way to use `packages`, we could extend it for `go.#Build` as well (since the go build command can accept a list of packages as well)

**Comment by aluzzardi** [comment]
Thank you @dolanor!

/cc @helderco -- should we make this backward compatible?

**Comment by helderco** [comment]
> BTW, what's our policy on breaking change? Anything that is public and usable by user that has changed is considered as breaking change? Be that CUE or Go? From what I know, actually Go packages are not really public in the first place, right?

Go isn't public, no. I put my thoughts in:
- https://github.com/dagger/dagger/issues/2251

**Comment by dolanor** [comment]
We decided I will keep backwards compatibility:
`package` as a string
`packages` as an array

File: pkg/universe.dagger.io/go/test/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue:13]: Maybe you're right. I wasn't sure if it was an improvement or just more confusion.
I guess it might just be more confusion.

**Comment by dolanor** [comment]
Finally got around getting the backward compatibilty working.

Would love help on how to factor
https://github.com/dagger/dagger/pull/2210/files#diff-357ac11ec1253d5a1c35c22f6e84931a14e2d614395ebc5300c48e1374049198R13-R25
and
https://github.com/dagger/dagger/pull/2210/files#diff-ad0a91c580ee3114755bc8f3456b110d484e7118f520647c8484b78f49e2de05R44-R56

File: pkg/universe.dagger.io/go/test/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue]: done. `packages` will always be a list

File: ci.cue
Comment (TomChv) [inline ci.cue:85]: Nit: Why not just `package`?

File: docs/plans/go-ci/plan.cue
Comment (TomChv) [inline docs/plans/go-ci/plan.cue:15]: Nit: Why not just package?

File: pkg/universe.dagger.io/go/build.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/build.cue:57]: Add a `FIXME: find a better workaround with disjunction`

Maybe someone will find it :cry:

File: pkg/universe.dagger.io/go/test.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/test.cue:25]: Add a `FIXME: find a better workaround with disjunction`

Maybe someone will find it :cry:

File: pkg/universe.dagger.io/go/test.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/test.cue:13]: Nit: `| []` should be useless

File: pkg/universe.dagger.io/go/build.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/build.cue:44]: Nit: `| []` should be useless

File: pkg/universe.dagger.io/go/test/test.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/test/test.cue:21]: Can yo test with both too please?

File: docs/plans/go-ci/plan.cue
Comment (dolanor) [reply docs/plans/go-ci/plan.cue:15]: Because it is now the deprecated type.

File: pkg/universe.dagger.io/go/build.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/build.cue:57]: Also, just if we could factor the one in the build.cue and test.cue, that would be good. I didn't succeed, though.

File: pkg/universe.dagger.io/go/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test.cue:13]: clearly. Some debugging code I didn't remove…

File: pkg/universe.dagger.io/go/build.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/build.cue:44]: clearly. Some debugging code I didn't remove…

File: pkg/universe.dagger.io/go/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test.cue:25]: Also, just if we could factor the one in the build.cue and test.cue, that would be good. I didn't succeed, though.

File: pkg/universe.dagger.io/go/test/build.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/build.cue:66]: nice! It simplifies it!

File: pkg/universe.dagger.io/go/test/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue:21]: Ok, I'll change the simple to use the default case.

File: pkg/universe.dagger.io/go/test/build.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/build.cue:47]: done

File: pkg/universe.dagger.io/go/test/build.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/build.cue:41]: done

File: pkg/universe.dagger.io/go/test/test.cue
Comment (TomChv) [reply pkg/universe.dagger.io/go/test/test.cue]: I want you to test all cases possible haha

File: pkg/universe.dagger.io/go/test/test.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/test/test.cue:25]: Mixing mix test

**Review by TomChv — APPROVED** [review]
LGTM! I left one last comment but we can take care of this later if you prefer.

File: pkg/universe.dagger.io/go/test/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue:25]: I don't understand your comment :stuck_out_tongue:

Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue:25]: I think I got it: `missing a test mixing package + packages`
You're right and I'll add this.

Comment (TomChv) [reply pkg/universe.dagger.io/go/test/test.cue:25]: Yes haha sorry

Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue:69]: Added it.
I tried to find a better way to do it, as I create temporary files for the right execution of each package under test.
Before each execution, my tests cleans up the FS.
I'd rather have it working without mount, but I don't know how to do that. I'll think it will be **good enough** for now.

**Review by TomChv — APPROVED** [review]
LGTM, rebase and fix CI and we merge it

**Comment by TomChv** [comment]
I rebase the PR myself, I also reuse `package: "./..."`syntax in `ci.cue` since definition is backward compatible.

I'll squash and merge that one after CI goes green.

**Comment by TomChv** [comment]
Tests are failing because `withBoth` is broken : https://github.com/dagger/dagger/runs/6379563005?check_suite_focus=true#step:8:358

**Local repo**
```
dagger do -p test.cue test withBoth --log-format plain
2:01AM INF actions.test.withBoth.test._image._dag."0"._op | computing
2:01AM INF client.filesystem."./data/hello".read | computing
2:01AM INF actions.test._baseImage._dag."0"._op | computing
2:01AM INF client.filesystem."./data/hello".read | completed    duration=0s
2:01AM INF actions.test._baseImage._dag."0"._op | completed    duration=0s
2:01AM INF actions.test.withBoth.verify._exec | computing
2:01AM ERR actions.test.withBoth.verify._exec | failed: process "sh -c test \"OK\" = \"`cat /src/greeting/greeting_test.result`\"\ntest \"OK\" = \"`cat /src/math/math_test.result`\"" did not complete successfully: exit code: 1
    duration=100ms
2:01AM ERR actions.test.withBoth.test._image._dag."0"._op | cancelled    duration=200ms
2:01AM INF actions.test.withBoth.verify._exec | #6 0.048 cat: can't open '/src/greeting/greeting_test.result': No such file or directory
2:01AM INF actions.test.withBoth.verify._exec | #6 0.048 cat: can't open '/src/math/math_test.result': No such file or directory
2:01AM FTL system | failed to execute plan: task failed: actions.test.withBoth.verify._exec: process "sh -c test \"OK\" = \"`cat /src/greeting/greeting_test.result`\"\ntest \"OK\" = \"`cat /src/math/math_test.result`\"" did not complete successfully: exit code: 1
```

File: pkg/universe.dagger.io/go/test/test.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/test/test.cue:43]: What is this doing? When I try locally nothing works

File: pkg/universe.dagger.io/go/test/test.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/test/test.cue:67]: What is this doing? When I try locally nothing works

File: pkg/universe.dagger.io/go/test/test.cue
Comment (TomChv) [inline pkg/universe.dagger.io/go/test/test.cue:52]: And since you take `_baseImage.output`, I do not understand how there could be `greeting_test.result`

File: pkg/universe.dagger.io/go/test/test.cue
Comment (TomChv) [reply pkg/universe.dagger.io/go/test/test.cue:69]: That currently do not work on my localhost, just executiong `go.#Test` should be enough no?

**Review by TomChv — CHANGES_REQUESTED** [review]
Tried locally and it do not works, I think `verify` is broken

**Comment by dolanor** [comment]
@TomChv and I peer coded on some solution to my problem: verify that each test was rightfully called.
Somehow my understanding of the FS export couldn't grasp how it functioned. The `.result` files would be saved in the container, but not exported, because they were on a mount.
We fixed this by fixing `go.#Container` so it used a `docker.#Copy`, but now, we have a disjunction between the `input` from the embedded `docker.#Run` and the `input` that we needed to create to make a copy. So for now, we stopped the embedding and put it behind a run key, but now other tests fail.

@dagger/maintainers: what should we do? Is the embedding of `docker.#Run` is important for its fields, and should we do some manual mapping to give access to it from the outside? Or do we just keep it behind a `run` field? Or is there some CUE magic that could help us do what we needed (`docker.#Copy` of the source FS instead of mounting it) without breaking everything else?

**Review by TomChv — APPROVED** [review]
`go.#Container` problem is solved by moving result file to `/tmp` so they are saved into the container filesystem.

This PR is ready to be merged :rocket:

**Comment by dolanor** [comment]
Forced push to fix some PGP signature that got wrong.

**Comment by dolanor** [comment]
@TomChv , you need to edit your signoff line. The DCO bot doesn't like it.
> Summary
>
>Commit sha: [4674fac](https://github.com/dagger/dagger/pull/2210/commits/4674fac4ff027a174dfdecdcb59f777f9c96de14), Author: Tom Chauveau, Committer: Tanguy ⧓ Herrmann; Expected "Tom Chauveau [tom.chauveau@epitech.eu](mailto:tom.chauveau@epitech.eu)", but got "Vasek - Tom C [tom.chauveau@epitech.eu](mailto:tom.chauveau@epitech.eu)".

File: pkg/universe.dagger.io/go/test/test.cue
Comment (helderco) [inline pkg/universe.dagger.io/go/test/test.cue:30]: This needs the `-e` flag, otherwise only last test will matter.

File: pkg/universe.dagger.io/go/test/test.cue
Comment (helderco) [inline pkg/universe.dagger.io/go/test/test.cue:49]: This needs the `-e` flag, otherwise only last test will matter.

File: pkg/universe.dagger.io/go/test.cue
Comment (helderco) [inline pkg/universe.dagger.io/go/test.cue:27]: I think this would be simpler:
```cue
package: *"." | string     // notice previous default preserved
packages: *[package] | [...string]   // previous behavior preservered as a default, allows overrides

#Container & {
    command:
        args: packages
        ...
}
```

This would mean that you wouldn't be able to use both at the same time, which is fine by me. The intent is to not break current configs, so if you're setting both `package` and `packages` you're updating to the new field, in which case I think it's ok to not make `package` work.

Ideally user should be warned if using both, but that will happen when we build the deprecation warnings. If we allow both to be used at the same time it also sets a precedent that that is a working config.

File: pkg/universe.dagger.io/go/test/test.cue
Comment (helderco) [inline pkg/universe.dagger.io/go/test/test.cue:69]: This needs the `-e` flag, otherwise only last test will matter.

**Review by helderco — COMMENTED** [review]
Just a suggestion, ok if you don't agree 🙂

File: pkg/universe.dagger.io/go/test/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue:30]: Good catch, thanks!

File: pkg/universe.dagger.io/go/test/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test/test.cue:69]: Good catch, thanks!

File: pkg/universe.dagger.io/go/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test.cue:27]: This was the reason this took so much time to fix.
And I thought about it enough to believe it is the right way.

## My suggestion:  `package` ∪ `packages` 

## Pros

- allow normal functioning even when mistakes are made by users (they didn't know `package` was being deprecated)
- function even if somewhere else in the package, somebody modify the `package(s)` fields

## Cons

- allow people to make mistake without thinking twice
- worse performances
- worse code/duplication

## Your suggestion: `package` Δ `packages`

## Pros

- less package code, cleaner
- better performance

## Cons

- user that make mistakes won't understand why they set `package` or `packages`, and it is ignored, unless they read our CUE package code. We might get more support.

**Comment by TomChv** [comment]
@dolanor What's the current status of the PR, are you waiting for @helderco response on your comment?

**Comment by mircubed** [comment]
@dolanor just checking in on the status of this PR. Let us know if you need anything

**Comment by mircubed** [comment]
@dolanor do you need anything else to update this PR for review?

File: pkg/universe.dagger.io/go/test.cue
Comment (dolanor) [reply pkg/universe.dagger.io/go/test.cue:27]: Can we have more opinions on which direction to take for this PR?

basically, my PR makes `package` + `packages` merge into a group of packages to test. But the code needed some if, otherwise we fell into disjunction land. So less performant.

@helderco thinks we should just have 1 option active at runtime, with preferences on `packages`, so if `package` and `packages` is set, test would ignore the former, and would run only on `packages`. User might be confused if they configured `package` and it is not tested at all.

Since it's the oldest opened PR, we could finally close it.

cc: @helderco @TomChv @aluzzardi @samalba @shykes

**Comment by dolanor** [comment]
> @dolanor do you need anything else to update this PR for review?

Just updated [here](https://github.com/dagger/dagger/pull/2210#discussion_r915859218)

File: pkg/universe.dagger.io/go/test.cue
Comment (gerhard) [reply pkg/universe.dagger.io/go/test.cue:27]: This looks straightforward to me:

```cue
package: *"." | string // notice previous default preserved packages: *[package] | [...string] // previous behavior preservered as a default, allows overrides
packages: *[package] | [...string] // previous behavior preservered as a default, allows overrides
```

Showing a deprecation warning at the top that does not get hidden by the default tty logger sounds like a nice way of eventually removing `package`. I think that it's OK for users that declare both to ask for support. I assume that there will not be more than a handful that do it before we remove `package`.

Comment (helderco) [reply pkg/universe.dagger.io/go/test.cue:27]: Yes, I prefer the deprecation option and then perhaps prioritize #2251 so that we can actually warn users about it.

```cue
// DEPRECATED: Use `packages` instead
package: *"." | string
packages: *[package] | [...string]
```

With that change, it'll become:

```cue
package: *"." | string @deprecated("use `packages` instead")
packages: *[package] | [...string]
```

And then we can detect on load if that field is used and trigger a warning with that message.

In any case, I'm also ok to merge as is and simplify/deprecate later when we have the warnings support. If so, please add a TODO for this.

Comment (gerhard) [reply pkg/universe.dagger.io/go/test.cue:27]: How do we resolve this one @dolanor?

Comment (dolanor) [reply pkg/universe.dagger.io/go/test.cue]: I like this deprecation notice. As talked with @gerhard, I'll already fix this PR with this soon to come update (it doesn't impact the plan, still works).
Also, we decided with @gerhard that finally, the package+packages use case might be too niche to bother and it might not create too many problems in the future (which is what I feared).
Also, the fix is **really** easy, `s/package/packages/`.

**Comment by dolanor** [comment]
Having many timeouts with GHA.
Is there a way we can fix that? (except providing our own infra offering :p)

**Comment by dolanor** [comment]
Ok, I think I know what was blocking me. I thought the `/tmp` was insulated from modification, but it seems that some of the test delete/overwrite files in parallel which creates conflict.

Using the same `dagger.#FS`, that is then used as a mount inside each action, and that those test from this FS write to `/tmp` (which is a normal folder in the action container), could any other action impact this `/tmp` folder?

My guess is, no. Is that correct?

**Comment by dolanor** [comment]
I don't understand how a file that has been created here:
https://github.com/dagger/dagger/runs/7500277333?check_suite_focus=true#step:8:1392

Doesn't exist in the next action that has the output directly in:
https://github.com/dagger/dagger/runs/7500277333?check_suite_focus=true#step:8:1453

Any @dagger/maintainers can help me figure out? It seems it's an issue with buildkit? Am I setting up things wrong?

**Comment by dolanor** [comment]
Did a session with @grouville , and I learned that to export something from a layer (from the mount in that case) into the `output` of a step, you need to merge it in the rootfs.
Any top layer doesn't get automatically exported (which was my assumption on how I interact with docker).
So I have to copy the result files into the image.

**Comment by gerhard** [comment]
Good job on pushing this forward, inch by inch. Almost there!

**Comment by gerhard** [comment]
Leaving this with you @dolanor @grouville. Will check again on Aug. 1.

**Comment by grouville** [comment]
@jpadams. I rebased all the branch and fixed the last things blocking. This PR is ready-to-be-merged. Do not hesitate to add comments

**Comment by grouville** [comment]
@jpadams @helderco, this can be reviewed one more time. I fixed the last remaining issues and rebased to current `latest`

File: pkg/universe.dagger.io/go/test/data/hello/math/algebra_test.go
Comment (dolanor) [inline pkg/universe.dagger.io/go/test/data/hello/math/algebra_test.go:16]: now that we don't use `os.CreateTemp()`, we don't need the `*` in the pattern.

File: pkg/universe.dagger.io/go/test/data/hello/internal/testutil/result_file.go
Comment (dolanor) [inline pkg/universe.dagger.io/go/test/data/hello/internal/testutil/result_file.go:30]: remove those comments

File: pkg/universe.dagger.io/go/build.cue
Comment (dolanor) [inline pkg/universe.dagger.io/go/build.cue:52]: use the same system as the one in go.#Test.
Using disjunction, and ignoring `package` if `packages` is set.

File: pkg/universe.dagger.io/go/test/build.cue
Comment (dolanor) [inline pkg/universe.dagger.io/go/test/build.cue:39]: need to add a `withBoth` that should behave the same as `go.#Test` tests.

**Review by dolanor — COMMENTED** [review]
Added some comments for improving.
I'll link it to a new issue.

**Pull request merged**
